sound/soc/raspberrypi/rp1_aout.c

```diff
@@ -0,0 +1,360 @@
+	val = FIELD_PREP_CONST(AUDIO_OUT_FIFO_CONTROL_DWELL_TIME, 2) |
+		FIELD_PREP_CONST(AUDIO_OUT_FIFO_CONTROL_THRESHOLD, 0x10) |
+		AUDIO_OUT_FIFO_CONTROL_DMA_DREQ_EN;
+	aout_reg_wr(aout, AUDIO_OUT_FIFO_CONTROL, val);
+}
+
+static void audio_startup(struct rp1_aout *aout)
+{
+	u32 val;
+
+	/* CIC rate 10, for an overall upsampling ratio of 40 */
+	val =  FIELD_PREP_CONST(AUDIO_OUT_CTRL_CIC_RATE, 0xa) |
+		AUDIO_OUT_CTRL_LEFT_CH_ENABLE | AUDIO_OUT_CTRL_RIGHT_CH_ENABLE;
+	aout_reg_wr(aout, AUDIO_OUT_CTRL, val);
+
+	/* Press the "go" button */
+	val |= AUDIO_OUT_CTRL_PERIPH_EN;
+	aout_reg_set(aout, AUDIO_OUT_CTRL, val);
+
+	/* Poke zeroes in to avoid undefined values on underrun */
+	aout_reg_wr(aout, AUDIO_OUT_SAMPLE_FIFO, 0);
+}
+
+static void audio_start_unmute(struct rp1_aout *aout)
+{
+	/* Set the mute engine to start unmuting */
+	aout_reg_clr(aout, AUDIO_OUT_MUTE_CTRL_LEFT,
+		     AUDIO_OUT_MUTE_CTRL_LR_INIT_UNMUTE |
+		     AUDIO_OUT_MUTE_CTRL_LR_INIT_MUTE);
+	aout_reg_clr(aout, AUDIO_OUT_MUTE_CTRL_RIGHT,
+		     AUDIO_OUT_MUTE_CTRL_LR_INIT_UNMUTE |
+		     AUDIO_OUT_MUTE_CTRL_LR_INIT_MUTE);
+	aout_reg_set(aout, AUDIO_OUT_MUTE_CTRL_LEFT,
+		     AUDIO_OUT_MUTE_CTRL_LR_INIT_UNMUTE);
+	aout_reg_set(aout, AUDIO_OUT_MUTE_CTRL_RIGHT,
+		     AUDIO_OUT_MUTE_CTRL_LR_INIT_UNMUTE);
+}
+
+static void audio_mute_sync(struct rp1_aout *aout)
+{
+	u32 mask = 0x1 | 0x4; /* stable states of the mute FSM */
+
+	while ((aout_reg_rd(aout, AUDIO_OUT_MUTE_CTRL_RIGHT) & mask) != 0)
+		usleep_range(1000, 5000);
+}
+
+/* Device DAI interface */
+
+static int rp1_aout_startup(struct snd_pcm_substream *substream,
+			    struct snd_soc_dai *dai)
+{
+	struct rp1_aout *aout = dev_get_drvdata(dai->dev);
+	struct snd_soc_pcm_runtime *rtd = snd_soc_substream_to_rtd(substream);
+	struct snd_soc_dai_link *dai_link = rtd->dai_link;
+
+	dai_link->trigger_stop = SND_SOC_TRIGGER_ORDER_LDC;
+
+	if (!aout->initted) {
+		dev_info(dai->dev, "RP1 Audio Out start\n");
+		audio_init(aout);
+		audio_startup(aout);
+		audio_start_unmute(aout);
+		aout->initted = true;
+	}
+
+	return 0;
+}
+
+static int rp1_aout_hw_params(struct snd_pcm_substream *substream,
+			      struct snd_pcm_hw_params *params,
+			      struct snd_soc_dai *dai)
+{
+	struct rp1_aout *aout = dev_get_drvdata(dai->dev);
+
+	/* We only support this one configuration */
+	if (params_rate(params) != 48000 || params_channels(params) != 2 ||
+	    params_format(params) != SNDRV_PCM_FORMAT_S16_LE) {
+		dev_err(dai->dev, "Invalid HW params\n");
+		return -EINVAL;
+	}
+
+	audio_mute_sync(aout);
+
+	return 0;
+}
+
+static int rp1_aout_trigger(struct snd_pcm_substream *substream, int cmd,
+			    struct snd_soc_dai *dai)
+{
+	struct rp1_aout *aout = snd_soc_dai_get_drvdata(dai);
+
+	if (cmd == SNDRV_PCM_TRIGGER_STOP ||
+	    cmd == SNDRV_PCM_TRIGGER_SUSPEND ||
+	    cmd == SNDRV_PCM_TRIGGER_PAUSE_PUSH) {
+		/* Push a zero sample (assuming DMA has stopped already?) */
+		aout_reg_wr(aout, AUDIO_OUT_SAMPLE_FIFO, 0);
```

Comment:
Underflow should be handled gracefully by the hardware (the last sample plays forever) - ASoC decouples DMA ops from driver visibility so you may introduce a glitch here if DMA is still running. I don't know if it synchronously drains the DMA buffers or flushes them.

Reply:
In tests it does seem  to use the given value quite reliably.

It wasn't clear to me whether a step to zero would be more or less objectionable than a nonstandard DC value (perhaps followed by a step later). Maybe 'stop' and 'pause' cases should be treated differently?

Reply:
After a bit more debug: I haven't actually observed a PAUSE_PUSH or SUSPEND -- when I Ctrl-Z `aplay`, it does a STOP. I think I will keep the zero as it seems to work, and may be better than sticking at an arbitrary value.